src/library/scala/collection/immutable/ListSet.scala

```diff
@@ -110,14 +110,37 @@ sealed class ListSet[A] extends AbstractSet[A]
             var toAdd = ls
             var result: ListSet[A] = this
 
+            var pending: Array[A] = null
+            var pending0, pending1, pending2, pending3: A = null.asInstanceOf[A]
+            var pendingCount = 0
             while (toAdd ne lsTail) {
               val elem = toAdd.elem
               if (!containsLimited(result, elem, lsTail)) {
-                val r = result
-                result = new r.Node(elem)
+                pendingCount match {
+                  case 0 => pending0 = elem
+                  case 1 => pending1 = elem
+                  case 2 => pending2 = elem
+                  case 3 => pending3 = elem
+                  case _ =>
+                    if (pending eq null) pending = new Array[AnyRef](lsSize - remaining - 4).asInstanceOf[Array[A]]
+                    pending(pendingCount -4) = elem
```

Comment:
nit: tiny formatting issue that's not super important

```suggestion
                    pending(pendingCount - 4) = elem
```

Reply:
done in reformat of override def ++`